Initial merge/PR of the skill oracle

robinsonkwame opened the pull request:
This is the initial merge/pr of the skill oracle. It's far ahead enough of the master that it should be brought in.

Unittest note:
unittest run with `python -m unittest` will give a ResourceWarning error on a unclosed socket and spit out `nc` usage information.
unittest run with `py.test --ipdb --fulltrace` will throw these warnings or `nc` usage information.

robinsonkwame committed:
inital commit, not all files refactored from copy

robinsonkwame committed:
Adds vowpal wabbit dockerfile for ubuntu 14, some renamed but not refactored test files

robinsonkwame committed:
Updates workdir in Dockerfile

robinsonkwame committed:
update webserver port

robinsonkwame committed:
getting ready to start on a main program to use vw, learn and provide
canditate skills when asked

robinsonkwame committed:
WIP: oracle, TDD, starting with daemon standing up and killing

robinsonkwame committed:
WIP: getting oracle daemonized, something odd with check_socket,
probably host=None is a bad thing for it, need to provide default
for case of checking when none was provided (assume local host basically)

robinsonkwame committed:
WIP: gives basic daemon stand up, tear down, unit test

robinsonkwame committed:
WIP: added more documentatoin to the file

robinsonkwame committed:
removes old test_

robinsonkwame committed:
refactors tests to take out teardown, standup

robinsonkwame committed:
WIP: tsting PUSH, can i push to vw directly with teach?

robinsonkwame committed:
WIP: initall pass of .PUT, need to put esacping of data in

robinsonkwame committed:
WIP: adds in escaping to PUT call, ... i assume it works as expected

robinsonkwame committed:
WIP: starting to stand up PUT endpoint, really just a pass through

robinsonkwame committed:
WIP: attempting to test webserver, getting some weird zombie socket that i
can't see, taking network down, up didnt' help, restarting

robinsonkwame committed:
WIP: circling back to debug/fix the stand up and tear down logic, it appeared to
work at first but apparently never did. killall will work for tearing down vw
but check_call() doesn't seem to capture vw's stdout when used with echo, nc
as a way to verify its running. Need to look into this more...

robinsonkwame committed:
Update test_skilloracle.py

WIP:   new check, by successfully tearing down the oracle we confirm  a) it started up and b) it was killed
 So, all we need to do is remove the assert True above and proceed directly with the below

robinsonkwame committed:
Update skilloracle.py

WIP: side steps check_socket which doesn't' seem to work as expected for me. todo: either fix it or remove it

robinsonkwame committed:
WIP: adds beefed up check of existence of oracle (namely that it replies
to an instance with some kind of response)

robinsonkwame committed:
WIP: placed new logic into unitttests, about to test out

robinsonkwame committed:
WIP: removes socket logic, doesn't seem to connect to vowpal wabbit even
a netcat from the shell will. See: https://github.com/requests/requests/issues/1882
for old but possible interactions that are going on here.

robinsonkwame committed:
WIP: should finish out PUT, standup/teardown unit tests, yay

robinsonkwame committed:
removes uneeded modules from unit test, skilloracle

robinsonkwame committed:
renamed a unittest

robinsonkwame committed:
WIP: tweaked docker-compose to reflect the skill oracle

robinsonkwame committed:
Prepare the skilloracle to easily merge into the master a simple directory.
Keeping everything in a directory preserves the structure of the service
while allowing it to uniquely boot up its own webserver, etc

robinsonkwame committed:
Container passes `py.test --pdb --fulltrace` unittests

robinsonkwame committed:
WIP: starts standard GET, need to add in pipeline, multi/exec
re: https://groups.google.com/forum/#!topic/redis-db/ur9U8o-Sko0

robinsonkwame committed:
WIP: simple get, doesn't get cardinality yet, refill datastore

robinsonkwame committed:
WIP: roughly added recvline into PUT for cases were we want a prediction (no label),
follow up on fetch_push_more() unittest, should get responses that we can forward
to redis ...

robinsonkwame committed:
WIP: push_once passes first iteration of unittest, basically
after a GET, we can call push_once defensively to refill the
candidate GET store as needed (only pushes once per period,
no matter how many times called).

todo: needs access to live/real db service

robinsonkwame committed:
WIP: updates endpoint, but we need:
* a fetcher for the real database (need its API)
* need to suss out what indices for GET return value to
break out into candidate, context, importance, etc

robinsonkwame committed:
WIP: uncommented out skipped unittests

thcrock commented on skilloracle/docker-compose.yml:
```diff
@@ -0,0 +1,10 @@
+# This file is a WIP
```
If this is to tie all the dockerfiles together, should it be in the top level instead of the skill-oracle directory?

robinsonkwame replied:
Agreed, although the oracle service does depend on the official docker `redis` image we'll have to figure out a good way to express that dependency w/in the skill oracle service

thcrock commented on skilloracle/skilloracle/skilloracle/skilloracle.py:
```diff
@@ -0,0 +1,193 @@
+import time
```
This deep of a hierarchy with every level named 'skilloracle' is confusing. At the very least, if this .py file is the only thing that can be imported from this level, we can probably just put the contents in the __init__.py file. But I think there's other stuff to unpack here. What are all the levels?

robinsonkwame replied:
Agreed, I'll move things around to reflect something more like this (as discussed in #mlops):

```
Skill-labeler/skilloracle
Skill-labeler/test #all unit tests for each microservice
Skill-labeler/skilloracle/api # see next comment
Skill-labeler/skilloracle/endpoint # micro service related
Skill-labeler/skilloracle/oracle # ml smarts
Skill-labeler/skilloracle/webserver.py, run.sh # stands up service, its API
```

robinsonkwame commented:
Good point, let's move it up a directory 

Sent from my iPhone

On Nov 9, 2017, at 3:48 PM, Tristan Crockett <notifications@github.com> wrote:

@thcrock commented on this pull request.

In skilloracle/docker-compose.yml:

> @@ -0,0 +1,10 @@
+# This file is a WIP
If this is to tie all the dockerfiles together, should it be in the top level instead of the skill-oracle directory?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

thcrock commented on skilloracle/skilloracle/skilloracle/skilloracle.py:
```diff
@@ -0,0 +1,193 @@
+        response = None
+        pipe = self.redis_db.pipeline()# runs w/in multi/exec, atomic, on Redis
+
+        pipe.zrange(self.SKILL_CANDIDATES,
```
I'm curious what these three commands are retrieving. You're running three commands, and then in the endpoint you grab three pieces of data (candidate, context, importance). Am I correct in assuming that each of these commands is producing one of those pieces of data?

Some comments inline describing the Redis data structures being used and how they're used would be helpful. You may have this in some documentation elsewhere, but inlining some would make the code more clear.

robinsonkwame committed:
WIP: re organized directory structure, revising docker-compose file
so that a particular service, and its dependencies, can be simply
started up

robinsonkwame committed:
WIP: modifies docker-compose to stand up the skilloracle service

robinsonkwame committed:
Adds big documentation to __get_redis, as mentioned in the PR comments

robinsonkwame committed:
WIP Fixing up skilloracle towork under an older debian, than the Ubuntu 16 host system
I might just switch the image over to Ubuntu 16 to avoid all this

robinsonkwame committed:
WIP: updates skill oracle image to Ubuntu 16, rebased to new OS
need to verify that spawning, shell behavior is as expected.

robinsonkwame committed:
WIP: incorporates some fixes to enable unittests to run again, fixes up
the PUT response object, now a dictionary

robinsonkwame committed:
WIP: did some endpoint modifications, realized there's an
architecturally simpler way to deal with the skill candidate store,
putting values on to it, etc that doesn't couple with the database
as much (basically, make PUT also push to the candidate store on
unlabelled put and implement a candidate store size endpoint indicating
how many items are int he store. If user wants more then they can
do more puts after reading the candidate store size)

robinsonkwame committed:
Container run py.test --pbd --fulltrace tests all pass

robinsonkwame committed:
Refactors, corrects, the PR to:
a) Re organize directory layout for simplicity
b) Re bases the Dockerfile to run on Ubuntu 16.04, gets around some
shell/terminal spawning issues causing some unittests to fail
c) Modifies the PUT API to push unlabelled examples to the candidate store,
simplifies the skilloracle API to just GET, PUT calls.

robinsonkwame commented:
@thcrock the above changes:

a) Re organizes directory layout for simplicity
b) Re bases the Dockerfile to run on Ubuntu 16.04, gets around some shell/terminal spawning issues causing some unittests to fail, creates a docker-compose skilloracle service
c) Modifies the PUT API to push unlabelled examples to the candidate store, simplifies the skilloracle API to just GET, PUT calls.
d) Adds additional comments for clarification
